src/dataflow/src/server.rs

```diff
@@ -900,7 +900,15 @@ where
                     } else {
                         true // retain
                     }
-                })
+                });
+
+                // Respond with cancellation.
+                self.feedback_tx
```

Comment:
If we now always expect exactly the right number of responses from dataflow workers, then the if let at https://github.com/MaterializeInc/materialize/blob/33adcc86052c5568bb5dd0e823729e11fdef3e7d/src/coord/src/coord.rs#L599 should get changed to a `.expect()` to formally enforce the contract between coord and dataflow.

Reply:
(And that comment should be changed/removed to explain.)

Light preference for the cancellation fix to be in its own PR because I think it's an existing bug. It also needs to come with a test. It's easily testable right now from main with:

```
materialize=> create table t (i int);
CREATE TABLE
materialize=> select * from t as of now()+'1h';
^CCancel request sent
```

which incorrectly blocks.

Reply:
Quick thought: if you change the code to an `expect()` you are locking in the behavior that it is not acceptable for the coordinator to respond to a cancellation without waiting for the dataflow responses. That seems to lock in potentially defective behavior (e.g. when a query is hung because `dataflow` is non-responsive). It sounds like what you want is that the dataflow layer should always explicitly confirm a cancellation, which seems sane, but that it different from requiring the coordinator to await that cancellation / crashing if it eagerly cancels the query.

Reply:
coord already relies on very specific behavior from dataflow, that each worker returns exactly one response. But dataflow workers were only sending a message if they weren't cancelled, which is just kinda annoying?

I agree that in the case of dataflow being non-responsive, waiting for a cancellation message could block forever/too long.

Is there a better way for coord to wait for all dataflow worker responses besides waiting for (# of dataflow workers) responses? That breaks if a worker incorrectly sends two or zero (which was this bug).

Maybe instead of waiting for N messages, we keep a hashmap with N entries, 1 per worker, whose key is the worker id. This would allow coord to detect duplicate messages (which would be bad as those could incorrectly decrement the waiting counter).

Reply:
I think it is fine to want the dataflow layer to produce exactly one message (per the above comment, this change doesn't do that yet, nor does it fix it yet for tail), but that is different from making the coordinator block awaiting all of these messages. At some level, coord can do what it likes; definitely start with what is easiest to verify. But we'll eventually (not this PR) want to get to a spot where the coordinator responds as soon as it has the information it needs.

Reply:
I've come around to this. I think the change should be to leave dataflow as is, have handle_cancel send exactly 1 cancel message on the channel, and then explicitly drop the pending_peek.

handle_cancel incorrectly thought it needed to send one cancel per worker, but the count thing only matters if those messages are received by `message_worker`, which they aren't, they are received by the `.fold` operator in `sequence_peek`, which is why 1 message is sufficient.